### PR TITLE
GH-36546: [Swift] The initial implementation for swift arrow flight

### LINKED ISSUE #36546: [Swift] Initial Swift Arrow Flight impl
### Describe the enhancement requested

The initial implementation for swift Arrow Flight.

### Component(s)

Swift

### PR DESCRIPTION
**Pull request by abandy**
The initial implementation for swift arrow flight.  

This change depends on PR GH-36544.
* Closes: #36546

### CODE REVIEW

**Comment by kou** [comment]
I've merged #36544. Could you rebase on main?

**Comment by abandy** [comment]
> I've merged #36544. Could you rebase on main?

Awesome, thank you!  Will do.

File: swift/ArrowFlight/Tests/ArrowFlightTests/FlightTest.swift
Comment (kou) [inline swift/ArrowFlight/Tests/ArrowFlightTests/FlightTest.swift:1]: Could you add a license header?

Comment (abandy) [reply swift/ArrowFlight/Tests/ArrowFlightTests/FlightTest.swift:1]: Good catch! Will do.

File: swift/ArrowFlight/Sources/ArrowFlight/FlightClient.swift [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    }
+    
+    public func listActions(_ closure: (FlightActionType) -> Void) async throws {
+        let listActions = client.makeListActionsCall(Arrow_Flight_Protocol_Empty())
+        for try await data in listActions.responseStream {
+            closure(FlightActionType(data))
+        }
+    }
+    
+    public func listFlights(_ criteria :FlightCriteria, closure: (FlightInfo) throws -> Void ) async throws {
```
Comment (kou) [inline swift/ArrowFlight/Sources/ArrowFlight/FlightClient.swift:38]:
```suggestion
    public func listFlights(_ criteria: FlightCriteria, closure: (FlightInfo) throws -> Void) async throws {
```

(We may want to introduce code formatter.)

File: swift/ArrowFlight/Sources/ArrowFlight/RecordBatchStreamReader.swift [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        if batchIndex < batches.count {
+            let batch = batches[batchIndex]
+            batchIndex += 1
+            return batch
+        }
+         
+        while true {
+            let flight_data = try await self.streamIterator.next()
```
Comment (kou) [inline swift/ArrowFlight/Sources/ArrowFlight/RecordBatchStreamReader.swift:47]: `flightData`?

File: swift/ArrowFlight/Tests/ArrowFlightTests/FlightTest.swift [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+    func doExchange(_ reader: ArrowFlight.RecordBatchStreamReader, writer: ArrowFlight.RecordBatchStreamWriter) async throws {
+        do {
+            for try await rb in reader {
+                XCTAssertEqual(rb.schema.fields.count, 3)
+                XCTAssertEqual(rb.length, 4)
+            }
+            
+            let rb = try makeRecordBatch()
+            try await writer.write(rb)
+        }catch {
```
Comment (kou) [inline swift/ArrowFlight/Tests/ArrowFlightTests/FlightTest.swift:80]:
```suggestion
        } catch {
```

File: swift/ArrowFlight/Tests/ArrowFlightTests/FlightTest.swift [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+            // Start the server and print its address once it has started.
+            FlightServerImpl.server = try await Server.insecure(group: group)
+                .withServiceProviders([provider])
+                .bind(host: "localhost", port: 8088)
+                .get()
+            
+            print("server started on port \(server!.channel.localAddress!.port!)")
+
+            // Wait on the server's `onClose` future to stop the program from exiting.
+        }catch {
```
Comment (kou) [inline swift/ArrowFlight/Tests/ArrowFlightTests/FlightTest.swift:145]:
```suggestion
        } catch {
```

File: swift/ArrowFlight/Tests/ArrowFlightTests/FlightTest.swift [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+            print("Unknown server error: \(error)")
+        }
+    }
+}
+
+public class FlightClientImpl {
```
Comment (kou) [inline swift/ArrowFlight/Tests/ArrowFlightTests/FlightTest.swift:151]: `FlightClientTester` or something may be better.

File: swift/ArrowFlight/Tests/ArrowFlightTests/FlightTest.swift [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+            num_call += 1
+            XCTAssertEqual(result.schema?.fields.count, 3)
+            XCTAssertEqual(result.batches[0].length, 4)
+        })
+        
+        XCTAssertEqual(num_call, 1)
+    }
+}
+
+var serverup = false
```
Comment (kou) [inline swift/ArrowFlight/Tests/ArrowFlightTests/FlightTest.swift:251]: Can we move this to `testFlightServer()`?

File: swift/ArrowFlight/Sources/ArrowFlight/FlightClient.swift (hunk 1 above)
Comment (abandy) [reply swift/ArrowFlight/Sources/ArrowFlight/FlightClient.swift:38]: Will change, agreed, I will start looking around.

File: swift/ArrowFlight/Sources/ArrowFlight/RecordBatchStreamReader.swift (hunk 2 above)
Comment (abandy) [reply swift/ArrowFlight/Sources/ArrowFlight/RecordBatchStreamReader.swift:47]: Will do.,

File: swift/ArrowFlight/Tests/ArrowFlightTests/FlightTest.swift (hunk 5 above)
Comment (abandy) [reply swift/ArrowFlight/Tests/ArrowFlightTests/FlightTest.swift:151]: Will change to FlightClientTester.

File: swift/ArrowFlight/Tests/ArrowFlightTests/FlightTest.swift (hunk 6 above)
Comment (abandy) [reply swift/ArrowFlight/Tests/ArrowFlightTests/FlightTest.swift:251]: Will move variable into a Actor.

File: swift/ArrowFlight/Tests/ArrowFlightTests/FlightTest.swift [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+            XCTAssertEqual(result.schema?.fields.count, 3)
+            XCTAssertEqual(result.batches[0].length, 4)
+        })
+        
+        XCTAssertEqual(num_call, 1)
+    }
+}
+
+actor FlightServerData {
+    public var serverup = false
```
Comment (kou) [inline swift/ArrowFlight/Tests/ArrowFlightTests/FlightTest.swift:252]: We may want to use `serverUp` for this too.

**Review by kou — APPROVED** [review]
+1

**Pull request merged**